File: src/modules/Pieces/Kinds/Pawn/Pawn.cpp
#pragma once

#include "Pawn.h"

#include "../../../Common/InstanceOf/InstanceOf.h"

Pawn::Pawn(PieceParams* params):
    Piece<RawPosition>(params)
{}

/**
 * En passant is a special pawn capture that can only occur
 * when these rules are met:
 * 
 * 1. The capturing pawn must have advanced exactly three
 * ranks to perform this move.
 * 2. The captured pawn must have moved two squares in one
 * move, landing right next to the capturing pawn.
 * 3. The en passant capture must be performed on the turn
 * immediately after the pawn being captured moves. If the
 * player does not capture en passant on that turn, they
 * no longer can do it later.
 */
bool Pawn::getEnPassant() {
    return this->enPassant;
}

/**
 * Updates en passant value determined by the en passant
 * rules.
 */
void Pawn::updateEnPassantState(int prevPosition, int newPosition) {
    if (!this->hasGoneEnPassant) {
        bool canGoEnPassant = abs(prevPosition - newPosition) == 2;

        if (canGoEnPassant) {
            this->enPassant = true;
            this->hasGoneEnPassant = true;
        }
    } else {
        this->enPassant = false;
    }
}

void Pawn::onMoveEnd(int column, int row) {
    RawPosition* rawPosition = this->pieceImpl->getRawPosition();
    GameController* gameController = this->pieceImpl->getGameController();

    this->updateEnPassantState(rawPosition->getRow(), row);

    gameController->gameControllerImpl->setPawnToIgnore(this);

    Piece::onMoveEnd(column, row);

    // // if pawn reached the end of the board, trigger a
    // // pawn promotion
    // if (rawPosition.row === 0 || rawPosition.row === 7) {
    //     this.controller.pieceSelectorModal.renderModal(this);
    // }
}

vector<RawPosition> Pawn::getPossibleMoves() {
    GameController* gameController = this->pieceImpl->getGameController();
    unique_ptr<GameState>& gameState = gameController->gameControllerImpl->getGameState();

    vector<RawPosition> moves = vector<RawPosition>();

    bool isWhite = this->pieceImpl->getColor() == PieceColor::WHITE;

    RawPosition* position = this->pieceImpl->getRawPosition();

    // Move forward
    int forward = (isWhite ? -1 : 1);

    int rowOfStart = (forward == 1 ? 1 : 6);

    int extraForward = (position->getRow() == rowOfStart
        ? forward
        : 0);

    int times = abs(forward + extraForward);

    for (int i = 1; i <= times; i++) {
        RawPosition nextPosition = RawPosition(
            position->getColumn(),
            position->getRow() + (i * forward)
        );

        GetValidPieceResponse response = gameState->gameStateImpl->getValidPiece(
            nextPosition.getColumn(),
            nextPosition.getRow()
        );

        if (response.isOutOfScope || InstanceOf::check<Piece<RawPosition>>(response.piece)) break;

        moves.push_back(nextPosition);
    }

    // Move diagonally to eat!
    RawPosition leftDiagonal  = RawPosition(position->getColumn() - 1, position->getRow() + forward);
    RawPosition rightDiagonal = RawPosition(position->getColumn() + 1, position->getRow() + forward);

    GetValidPieceResponse leftResponse  = gameState->gameStateImpl->getValidPiece(
        leftDiagonal.getColumn(),
        leftDiagonal.getRow()
    );
    GetValidPieceResponse rightResponse = gameState->gameStateImpl->getValidPiece(
        rightDiagonal.getColumn(),
        rightDiagonal.getRow()
    );

    if (
        !leftResponse.isOutOfScope
        && InstanceOf::check<Piece<RawPosition>>(leftResponse.piece)
        && leftResponse.piece->pieceImpl->getColor() != this->pieceImpl->getColor()
    ) {
        moves.push_back(leftDiagonal);
    }
    if (
        !rightResponse.isOutOfScope
        && InstanceOf::check<Piece<RawPosition>>(rightResponse.piece)
        && rightResponse.piece->pieceImpl->getColor() != this->pieceImpl->getColor()
    ) {
        moves.push_back(rightDiagonal);
    }

    // En Passant
    GetValidPieceResponse leftPawnResponse = gameState->gameStateImpl->getValidPiece(
        position->getColumn() - 1,
        position->getRow()
    );
    GetValidPieceResponse rightPawnResponse = gameState->gameStateImpl->getValidPiece(
        position->getColumn() + 1,
        position->getRow()
    );

    if (
        !leftPawnResponse.isOutOfScope
        && InstanceOf::check<Pawn>(leftPawnResponse.piece)
        && leftPawnResponse.piece->pieceImpl->getColor() != this->pieceImpl->getColor()
        && static_cast<Pawn*>(leftPawnResponse.piece)->getEnPassant()
    ) {
        RawPosition rawPosition = RawPosition();

        rawPosition.setRow(position->getRow() + forward);
        rawPosition.setColumn(position->getColumn() - 1);

        moves.push_back(rawPosition);
    }
    if (
        !rightPawnResponse.isOutOfScope
        && InstanceOf::check<Pawn>(rightPawnResponse.piece)
        && rightPawnResponse.piece->pieceImpl->getColor() != this->pieceImpl->getColor()
        && static_cast<Pawn*>(rightPawnResponse.piece)->getEnPassant()
    ) {
        RawPosition rawPosition = RawPosition();

        rawPosition.setRow(position->getRow() + forward);
        rawPosition.setColumn(position->getColumn() + 1);

        moves.push_back(rawPosition);
    }

    return moves;
}

File: packages/interactive_2d/chess/chess_react/src/Sketches/ChessSketch/elements/Pieces/Kinds/Pawn.ts
import { Piece } from "../Piece";
import { Position } from "../types";

export class Pawn extends Piece {
    private _enPassant = false;
    private _hasGoneEnPassant = false;

    /**
     * En passant is a special pawn capture that can only occur
     * when these rules are met:
     * 
     * 1. The capturing pawn must have advanced exactly three
     * ranks to perform this move.
     * 2. The captured pawn must have moved two squares in one
     * move, landing right next to the capturing pawn.
     * 3. The en passant capture must be performed on the turn
     * immediately after the pawn being captured moves. If the
     * player does not capture en passant on that turn, they
     * no longer can do it later.
     */
    public get enPassant(): boolean {
        return this._enPassant;
    }

    /**
     * Updates en passant value determined by the en passant
     * rules.
     */
    public updateEnPassantState(prevPosition?: number, newPosition?: number): void {
        if (!this._hasGoneEnPassant) {
            const canGoEnPassant: boolean = Math.abs((prevPosition ?? 0) - (newPosition ?? 0)) === 2;

            if (canGoEnPassant) {
                this._enPassant = true;
                this._hasGoneEnPassant = true;
            }
        } else {
            this._enPassant = false;
        }
    }

    /**
     * @override
     */
    public getPossibleMoves(): Position[] {
        const { board } = this.controller.gameState;

        console.log({ board, location : "getPossibleMoves" });

        const moves: Position[] = [];
        const isWhite = this.color === "white";
        const position = this.position;

        // Move forward
        const forward = isWhite ? -1 : 1;

        const rowOfStart = forward === 1 ? 1 : 6;

        const extraForward = position.row === rowOfStart
            ? forward
            : 0;

        const times = Math.abs(forward + extraForward);

        for (let i = 1; i <= times; i++) {
            const nextPosition: Position = {
                row    : position.row + (i * forward), 
                column : position.column,
            };
            const nextPiece: Piece | null | undefined = board?.[nextPosition.row]?.[nextPosition.column];
    
            if (typeof nextPiece === "undefined" || nextPiece instanceof Piece) break;

            moves.push(nextPosition);
        }

        // Move diagonally to eat!
        const leftDiagonal: Position = {
            row    : position.row + forward,
            column : position.column - 1,
        };
        const rightDiaonal: Position = {
            row    : position.row + forward,
            column : position.column + 1,
        };

        const leftPiece: Piece | null = board?.[leftDiagonal.row]?.[leftDiagonal.column] ?? null;
        const rightPiece: Piece | null = board?.[rightDiaonal.row]?.[rightDiaonal.column] ?? null;

        if (leftPiece instanceof Piece && leftPiece.color !== this.color) {
            moves.push(leftDiagonal);
        }
        if (rightPiece instanceof Piece && rightPiece.color !== this.color) {
            moves.push(rightDiaonal);
        }
        
        // En Passant
        const leftPawn: Piece | null = board?.[position.row]?.[position.column - 1] ?? null;
        const rightPawn: Piece | null = board?.[position.row]?.[position.column + 1] ?? null;

        if (
            leftPawn instanceof Pawn
            && leftPawn.color !== this.color
            && leftPawn.enPassant
        ) {
            moves.push({
                row    : position.row + forward,
                column : position.column - 1,
            });
        }
        if (
            rightPawn instanceof Pawn
            && rightPawn.color !== this.color
            && rightPawn.enPassant
        ) {
            moves.push({
                row    : position.row + forward,
                column : position.column + 1,
            });
        }

        return moves;
    }

    /**
     * @override
     */
    public onMoveEnd(rawPosition: Position): void {
        const previousPosition: Position = { ...this.position };
        
        this.updateEnPassantState(previousPosition.row, rawPosition.row);

        this.controller.pawnToIgnore = this;

        super.onMoveEnd(rawPosition);

        // if pawn reached the end of the board, trigger a
        // pawn promotion
        if (rawPosition.row === 0 || rawPosition.row === 7) {
            this.controller.pieceSelectorModal.renderModal(this);
        }
    }
}
